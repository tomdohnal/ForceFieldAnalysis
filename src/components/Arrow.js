// @flow
import React from 'react';

import { Box } from './common';
import { Rectangle, Triangle } from './svg';
import TextBubble from './TextBubble';

type Props = {
  left?: boolean,
  color: string,
  strength: number,
  name: string,
}

const Arrow = ({
  left, color, strength, name,
}: Props) => {
  const arrowWidth = strength * 50;
  const arrowHeight = 100;
  const rectangleHeight = arrowHeight - 40;
  const triangleWidth = 36;

  return (
    <Box>
      <TextBubble color={color} left={left}>{name}</TextBubble>
      <Box>
        <svg height={arrowHeight} width={arrowWidth} viewBox={`0 0 ${arrowWidth} ${arrowHeight}`}>
          <Rectangle
            width={arrowWidth - triangleWidth}
            height={rectangleHeight}
            color={color}
            x={left ? triangleWidth : 0}
            y={(arrowHeight - rectangleHeight) / 2}
          />
          <Triangle
            width={triangleWidth}
            height={arrowHeight}
            x={`${left ? triangleWidth : arrowWidth - triangleWidth},0`}
            y={`${left ? triangleWidth : arrowWidth - triangleWidth},${arrowHeight}`}
            z={`${left ? 0 : arrowWidth},${arrowHeight / 2}`}
            color={color}
          />
        </svg>
      </Box>
    </Box>
  );
};

export default Arrow;
