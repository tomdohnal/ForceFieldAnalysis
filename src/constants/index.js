export const MIN_STRENGTH = 1;
export const MAX_STRENGTH = 8;
