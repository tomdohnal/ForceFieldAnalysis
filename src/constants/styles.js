export const HEADER_HEIGHT = '60px';
export const FOOTER_HEIGHT = '80px';

export const COLORS = {
  WHITE: '#fff',
  BLACK: '#000',
  BLUE: '#268bd2',
  RED: '#dc2f2f',
  GREEN: '#859900',
};
