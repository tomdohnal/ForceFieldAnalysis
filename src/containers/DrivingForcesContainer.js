// @flow
import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Input, Button } from 'semantic-ui-react';

import DrivingForceArrow from '../components/DrivingForceArrow';
import { addDrivingForce, type Forces } from '../ducks/forces';

type State = {
  newDrivingForceName: string,
  newDrivingForceError: boolean,
  newDrivingForceErrorText: string,
};

type Props = {
  forces: Forces,
  addDrivingForce: (name: string, strength: number) => void,
};

class DrivingForcesContainer extends Component<Props, State> {
  state = {
    newDrivingForceName: '',
    newDrivingForceError: false,
    newDrivingForceErrorText: '',
  };

  onNewDrivingForceInputChange = (e) => {
    this.setState({
      newDrivingForceName: e.target.value,
      newDrivingForceError: false,
      newDrivingForceErrorText: '',
    });
  };

  onNewDrivingForceButtonClick = () => {
    if (!this.state.newDrivingForceName) {
      this.setState({
        newDrivingForceError: true,
        newDrivingForceErrorText: 'Enter the description for the force.',
      });
    } else {
      this.setState({
        newDrivingForceName: '',
      });

      this.props.addDrivingForce(this.state.newDrivingForceName, 2);
    }
  };

  renderForces() {
    return this.props.forces.map(force => (
      <DrivingForceArrow strength={force.strength} key={force.id} />));
  }

  render() {
    return (
      <div>
        {this.renderForces()}
        <Input
          onChange={this.onNewDrivingForceInputChange}
          value={this.state.newDrivingForceName}
          error={this.state.newDrivingForceError}
        />
        <Button onClick={this.onNewDrivingForceButtonClick}>New driving force</Button>
        <span>{this.state.newDrivingForceErrorText}</span>
      </div>
    );
  }
}

const mapStateToProps = state => ({
  forces: state.forces.filter(force => force.driving === true),
});

export default connect(mapStateToProps, { addDrivingForce })(DrivingForcesContainer);
