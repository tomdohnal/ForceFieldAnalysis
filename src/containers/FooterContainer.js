// @flow
import React from 'react';
import _ from 'lodash';
import { connect } from 'react-redux';

import { type Force } from '../ducks/forces';
import { type ReduxState } from '../redux';
import Footer from '../components/Footer';

type Props = {
  drivingForcesCount: number,
  hinderingForcesCount: number,
}

export const FooterContainer = ({
  drivingForcesCount, hinderingForcesCount,
}: Props) => {
  let centerText: string = '';
  if (drivingForcesCount === 0 && hinderingForcesCount === 0) {
    centerText = 'Create some forces and start the analysis!';
  } else if (drivingForcesCount === hinderingForcesCount) {
    centerText = 'It\'s a tie!';
  } else if (drivingForcesCount > hinderingForcesCount) {
    centerText = 'Let\'s make the change happen!';
  } else {
    centerText = 'No way! Keep status quo...';
  }

  return (
    <Footer
      drivingForcesCount={drivingForcesCount}
      hinderingForcesCount={hinderingForcesCount}
      centerText={centerText}
    />
  );
};


export const mapStateToProps = (state: ReduxState) => {
  const countForcesTotal = (driving: boolean) => (total: number, force: Force) => {
    if (force.driving === driving) {
      return total + force.strength;
    }

    return total;
  };

  return {
    drivingForcesCount: _.size(state.forces) ?
      _.reduce(state.forces, countForcesTotal(true), 0) : 0,
    hinderingForcesCount: _.size(state.forces) ?
      _.reduce(state.forces, countForcesTotal(false), 0) : 0,
  };
};

export default connect(mapStateToProps)(FooterContainer);
