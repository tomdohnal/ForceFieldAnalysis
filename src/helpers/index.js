// @flow
type Action = {
  type: string,
  payload?: any,
}

export type Reducer<S, A: Action> = (S, A) => S;

export type ReduxHandlers<S, A: Action> = {
  [key: string]: Reducer<S, A>,
}

export const createReducer =
  <S, A: Action>(initialState: S, handlers: ReduxHandlers<S, A>) =>
    (state: S = initialState, action: A) => (
      Object.prototype.hasOwnProperty.call(handlers, action.type) ?
        handlers[action.type](state, action) : state
    );

export const moveCaretAtTheEnd = (e: SyntheticEvent<HTMLInputElement>) => {
  if (e.target instanceof HTMLInputElement) {
    const { value } = e.target;
    e.target.value = '';
    e.target.value = value;
  }
};
