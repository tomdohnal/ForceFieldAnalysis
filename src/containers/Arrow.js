// @flow
import React, { Component } from 'react';
import { Form, Input, Button, Icon, Divider } from 'semantic-ui-react';
import EditIcon from 'react-icons/lib/ti/edit';
import { connect } from 'react-redux';

import { MAX_STRENGTH, MIN_STRENGTH } from '../constants/index';
import { COLORS } from '../constants/styles';
import { type Force, deleteForce, setForceName, increaseForce, decreaseForce } from '../ducks/forces';
import { Text, Box } from '../components/common/index';
import { Rectangle, Triangle } from '../components/svg/index';
import TextBubble from '../components/TextBubble';

type Props = {
  color: string,
  force: Force,
  deleteForce: (id: string) => void,
  setForceName: (id: string, name: string) => void,
  increaseForce: (id: string) => void,
  decreaseForce: (id: string) => void,
}

type State = {
  editArrow: boolean,
}

class Arrow extends Component<Props, State> {
  state = {
    editArrow: false,
  };

  onDeleteArrowIconClick = () => {
    this.props.deleteForce(this.props.force.id);
  };

  onIncreaseArrowIconClick = () => {
    if (this.props.force.strength !== MAX_STRENGTH) {
      this.props.increaseForce(this.props.force.id);
    }
  };

  onDecreaseArrowIconClick = () => {
    if (this.props.force.strength !== MIN_STRENGTH) {
      this.props.decreaseForce(this.props.force.id);
    }
  };

  onEditArrowInputChange = (event, data) => {
    this.props.setForceName(this.props.force.id, data.value);
  };

  onEditArrowNameClick = () => {
    this.setState({ editArrow: true });
  };

  onEditArrowNameFormSubmit = () => {
    this.setState({ editArrow: false });
  };

  render() {
    const { force: { driving, strength, name }, color } = this.props;

    const arrowUnitWidth = 50;
    const arrowWidth = strength * arrowUnitWidth;
    const arrowHeight = 80;
    const rectangleHeight = arrowHeight - 40;
    const triangleWidth = 36;

    return (
      <Box>
        <TextBubble color={color} left={!driving}>
          {this.state.editArrow ?
            <Form onSubmit={this.onEditArrowNameFormSubmit}>
              <Input autoFocus size="small" value={name} onChange={this.onEditArrowInputChange} action>
                <input style={{ width: 'inherit' }} />
                <Button size="small">Save changes</Button>
              </Input>
            </Form>
            :
            <Box>
              <Text fontSize="15px">{name}</Text>
              <EditIcon
                size="18"
                style={{ verticalAlign: 'top', marginLeft: '4px', cursor: 'pointer' }}
                onClick={this.onEditArrowNameClick}
              />
            </Box>
          }
        </TextBubble>
        <Box position="relative">
          <Icon
            name="delete"
            size="big"
            style={{
              position: 'absolute',
              top: 'calc(50% - 14px)',
              color: COLORS.RED,
              cursor: 'pointer',
              [driving ? 'right' : 'left']: `${arrowWidth}px`,
            }}
            onClick={this.onDeleteArrowIconClick}
          />
          <svg height={arrowHeight} width={arrowWidth} viewBox={`0 0 ${arrowWidth} ${arrowHeight}`}>
            <Rectangle
              width={arrowWidth - triangleWidth}
              height={rectangleHeight}
              color={color}
              x={!driving ? triangleWidth : 0}
              y={(arrowHeight - rectangleHeight) / 2}
            />
            <Triangle
              width={triangleWidth}
              height={arrowHeight}
              x={`${!driving ? triangleWidth : arrowWidth - triangleWidth},0`}
              y={`${!driving ? triangleWidth : arrowWidth - triangleWidth},${arrowHeight}`}
              z={`${!driving ? 0 : arrowWidth},${arrowHeight / 2}`}
              color={color}
            />
          </svg>
          <Box
            position="absolute"
            top={`${arrowHeight - ((arrowHeight - rectangleHeight) / 2)}px`}
            right={driving && `${triangleWidth}px`}
            left={!driving && `${triangleWidth}px`}
          >
            <Icon
              name="plus"
              size="large"
              disabled={strength === MAX_STRENGTH}
              style={{ cursor: `${strength < MAX_STRENGTH ? 'pointer' : 'initial'}` }}
              onClick={this.onIncreaseArrowIconClick}
            />
            <Text fontSize="18px">{strength}</Text>
            <Icon
              name="minus"
              size="large"
              disabled={strength === MIN_STRENGTH}
              style={{ cursor: `${strength > MIN_STRENGTH ? 'pointer' : 'initial'}` }}
              onClick={this.onDecreaseArrowIconClick}
            />
          </Box>
        </Box>
        <Divider hidden />
      </Box>
    );
  }
}

const actionCreators = { deleteForce, setForceName, increaseForce, decreaseForce };

export default connect(undefined, actionCreators)(Arrow);
