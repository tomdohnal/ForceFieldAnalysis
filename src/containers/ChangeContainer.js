// @flow
import React, { Component } from 'react';
import { connect } from 'react-redux';
import EditIcon from 'react-icons/lib/ti/edit';

import { COLORS } from '../constants/styles';
import type { ReduxState } from '../redux/index';
import { setChangeName, setChangeDescription, type Change } from '../ducks/change';
import { Box, Text } from '../components/common/index';
import ChangeNameForm from '../components/ChangeNameForm';
import ChangeDescriptionForm from '../components/ChangeDescriptionForm';

type Props = {
  change: Change,
  setChangeName: (name: string) => void,
  setChangeDescription: (description: string) => void,
};

type State = {
  editChangeName: boolean,
  editChangeDescription: boolean,
};

export class ChangeContainer extends Component<Props, State> {
  state = {
    editChangeName: !this.props.change.name,
    editChangeDescription: !this.props.change.description,
  };

  onNameFormSubmit = (): void => {
    this.setState({ editChangeName: false });
  };

  onEditNameClick = (): void => {
    this.setState({ editChangeName: true });
  };

  onDescriptionFormSubmit = (): void => {
    this.setState({ editChangeDescription: false });
  };

  onEditDescriptionClick = (): void => {
    this.setState({ editChangeDescription: true });
  };

  onNameChange = (e: SyntheticEvent<HTMLInputElement>, { value }: { value: string }): void => {
    this.props.setChangeName(value);
  };

  onDescriptionChange =
    (e: SyntheticEvent<HTMLInputElement>, { value }: { value: string }): void => {
      this.props.setChangeDescription(value);
    };

  render() {
    const { change: { name, description } } = this.props;

    return (
      <Box padding="20px 16px" backgroundColor={COLORS.BLUE} height="100%">
        {this.state.editChangeName ?
          <ChangeNameForm
            onSubmit={this.onNameFormSubmit}
            onInputChange={this.onNameChange}
            inputValue={name || ''}
          />
          :
          <Box marginTop="16px" textAlign="center">
            <Text color={COLORS.WHITE} fontSize="48px">{name}</Text>
            <Box display="inline" >
              <EditIcon
                color={COLORS.WHITE}
                size="28"
                style={{ verticalAlign: 'top', marginLeft: '8px', cursor: 'pointer' }}
                onClick={this.onEditNameClick}
              />
            </Box>
          </Box>
        }
        {this.state.editChangeDescription ?
          <Box marginTop="16px">
            <ChangeDescriptionForm
              onSubmit={this.onDescriptionFormSubmit}
              onInputChange={this.onDescriptionChange}
              inputValue={description || ''}
            />
          </Box>
          :
          <Box marginTop="16px">
            <Text color={COLORS.WHITE} fontSize="18px">{description}</Text>
            <Box display="inline" >
              <EditIcon
                color={COLORS.WHITE}
                size="18"
                style={{ verticalAlign: 'top', marginLeft: '4px', cursor: 'pointer' }}
                onClick={this.onEditDescriptionClick}
              />
            </Box>
          </Box>
        }
      </Box>
    );
  }
}

export const mapStateToProps = (state: ReduxState) => ({
  change: state.change,
});

export default connect(mapStateToProps, { setChangeName, setChangeDescription })(ChangeContainer);
