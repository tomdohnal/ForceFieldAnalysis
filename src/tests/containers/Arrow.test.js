import React from 'react';
import { shallow } from 'enzyme';

import { Arrow } from '../../containers/Arrow';
import { mockShallowComponentMethod } from '../helpers';
import { MAX_STRENGTH, MIN_STRENGTH } from '../../constants';

describe('Arrow', () => {
  const deleteForce = jest.fn();
  const setForceName = jest.fn();
  const increaseForce = jest.fn();
  const decreaseForce = jest.fn();

  const ShallowArrow = shallow(
    <Arrow
      color="#FFF"
      force={{
        id: '123',
        driving: true,
        strength: 5,
        name: 'the name of the force',
      }}
      deleteForce={deleteForce}
      setForceName={setForceName}
      increaseForce={increaseForce}
      decreaseForce={decreaseForce}
    />,
  );

  describe('rendering', () => {
    it('renders driving arrow correctly', () => {
      expect(ShallowArrow).toMatchSnapshot();
    });

    it('renders hindering arrow correctly', () => {
      const ShallowArrow = shallow(
        <Arrow
          color="#FFF"
          force={{
            id: '123',
            driving: true,
            strength: 5,
            name: 'the name of the force',
          }}
          deleteForce={deleteForce}
          setForceName={setForceName}
          increaseForce={increaseForce}
          decreaseForce={decreaseForce}
        />,
      );

      expect(ShallowArrow).toMatchSnapshot();
    });
  });

  describe('methods', () => {
    it('calls deleteForce action creator when onDeleteArrowIconClick is called', () => {
      ShallowArrow.instance().onDeleteArrowIconClick();

      expect(deleteForce).toBeCalled();
    });

    it('doesn\'t call increaseForce action creator when onIncreaseArrowIconClick is called and it would exceed the max strength', () => {
      ShallowArrow.instance().props.force.strength = MAX_STRENGTH;
      ShallowArrow.instance().onIncreaseArrowIconClick();

      expect(increaseForce).toHaveBeenCalledTimes(0);
    });

    it('calls increaseForce action creator when onIncreaseArrowIconClick is called', () => {
      ShallowArrow.instance().props.force.strength = MIN_STRENGTH;
      ShallowArrow.instance().onIncreaseArrowIconClick();

      expect(increaseForce).toBeCalled();
    });

    it('doesn\'t call decreaseForce action creator when onDecreaseArrowIconClick is called and it would exceed the max strength', () => {
      ShallowArrow.instance().props.force.strength = MIN_STRENGTH;
      ShallowArrow.instance().onIncreaseArrowIconClick();

      expect(decreaseForce).toHaveBeenCalledTimes(0);
    });

    it('calls decreaseForce action creator when onDecreaseArrowIconClick is called', () => {
      ShallowArrow.instance().props.force.strength = MAX_STRENGTH;
      ShallowArrow.instance().onIncreaseArrowIconClick();

      expect(increaseForce).toBeCalled();
    });

    it('calls setForceName action creator when onEditArrowInputChange is called', () => {
      ShallowArrow.instance().onEditArrowInputChange({}, { value: 'some some' });

      expect(setForceName).toBeCalled();
    });

    it('sets edit mode to true when onEditArrowNameClick is called', () => {
      ShallowArrow.instance().state.editArrow = false;

      ShallowArrow.instance().onEditArrowNameClick();

      expect(ShallowArrow.state().editArrow).toBeTruthy();
    });

    it('sets edit mode to false when onEditArrowNameFormSubmit is called', () => {
      ShallowArrow.instance().state.editArrow = true;

      ShallowArrow.instance().onEditArrowNameFormSubmit();

      expect(ShallowArrow.state().editArrow).toBeFalsy();
    });
  });

  describe('method triggering', () => {
    it('calls onEditArrowNameFormSubmit when the form is submitted', () => {
      // in order for the form to show, we have to set component's state and update it
      ShallowArrow.instance().state.editArrow = true;
      ShallowArrow.instance().forceUpdate();
      ShallowArrow.update();

      mockShallowComponentMethod(ShallowArrow, 'onEditArrowNameFormSubmit');

      ShallowArrow.find('Form').dive().find('form').simulate('submit');

      expect(ShallowArrow.instance().onEditArrowNameFormSubmit).toBeCalled();
    });

    it('calls onEditArrowInputChange when the form input is changed', () => {
      // in order for the form to show, we have to set component's state and update it
      ShallowArrow.instance().state.editArrow = true;
      ShallowArrow.instance().forceUpdate();
      ShallowArrow.update();

      mockShallowComponentMethod(ShallowArrow, 'onEditArrowInputChange');

      ShallowArrow.find('Input').dive().find('input').simulate('change');

      expect(ShallowArrow.instance().onEditArrowInputChange).toBeCalled();
    });

    it('calls onEditArrowNameClick when the edit arrow icon is clicked', () => {
      // in order for the form NOT to show, we have to set component's state and update it
      ShallowArrow.instance().state.editArrow = false;
      ShallowArrow.instance().forceUpdate();
      ShallowArrow.update();

      mockShallowComponentMethod(ShallowArrow, 'onEditArrowNameClick');

      ShallowArrow.find('TiEdit').dive().find('IconBase').dive().find('svg').simulate('click');

      expect(ShallowArrow.instance().onEditArrowNameClick).toBeCalled();
    });

    it('calls onDeleteArrowIconClick when the delete icon is clicked', () => {
      mockShallowComponentMethod(ShallowArrow, 'onDeleteArrowIconClick');

      ShallowArrow.find('Icon[name="delete"]').dive().find('i').simulate('click');

      expect(ShallowArrow.instance().onDeleteArrowIconClick).toBeCalled();
    });

    it('calls onIncreaseArrowIconClick when the increase icon is clicked', () => {
      mockShallowComponentMethod(ShallowArrow, 'onIncreaseArrowIconClick');

      ShallowArrow.find('Icon[name="plus"]').dive().find('i').simulate('click');

      expect(ShallowArrow.instance().onIncreaseArrowIconClick).toBeCalled();
    });

    it('calls onDecreaseArrowIconClick when the decrease icon is clicked', () => {
      mockShallowComponentMethod(ShallowArrow, 'onDecreaseArrowIconClick');

      ShallowArrow.find('Icon[name="minus"]').dive().find('i').simulate('click');

      expect(ShallowArrow.instance().onDecreaseArrowIconClick).toBeCalled();
    });
  });
});
