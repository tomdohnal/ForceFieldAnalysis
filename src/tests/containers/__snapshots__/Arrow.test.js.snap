// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Arrow rendering renders driving arrow correctly 1`] = `
<boxStyles>
  <textBubbleStyles
    color="#FFF"
    left={false}
  >
    <boxStyles>
      <text
        fontSize="15px"
      >
        the name of the force
      </text>
      <TiEdit
        onClick={[Function]}
        size="18"
        style={
          Object {
            "cursor": "pointer",
            "marginLeft": "4px",
            "verticalAlign": "top",
          }
        }
      />
    </boxStyles>
  </textBubbleStyles>
  <boxStyles
    position="relative"
  >
    <Icon
      as="i"
      name="delete"
      onClick={[Function]}
      size="big"
      style={
        Object {
          "color": "#dc2f2f",
          "cursor": "pointer",
          "position": "absolute",
          "right": "250px",
          "top": "calc(50% - 14px)",
        }
      }
    />
    <svg
      height={80}
      viewBox="0 0 250 80"
      width={250}
    >
      <Rectangle
        color="#FFF"
        height={40}
        width={214}
        x={0}
        y={20}
      />
      <Triangle
        color="#FFF"
        height={80}
        width={36}
        x="214,0"
        y="214,80"
        z="250,40"
      />
    </svg>
    <boxStyles
      left={false}
      position="absolute"
      right="36px"
      top="60px"
    >
      <Icon
        as="i"
        disabled={false}
        name="plus"
        onClick={[Function]}
        size="large"
        style={
          Object {
            "cursor": "pointer",
          }
        }
      />
      <text
        fontSize="18px"
      >
        5
      </text>
      <Icon
        as="i"
        disabled={false}
        name="minus"
        onClick={[Function]}
        size="large"
        style={
          Object {
            "cursor": "pointer",
          }
        }
      />
    </boxStyles>
  </boxStyles>
  <Divider
    hidden={true}
  />
</boxStyles>
`;

exports[`Arrow rendering renders hindering arrow correctly 1`] = `
<boxStyles>
  <textBubbleStyles
    color="#FFF"
    left={false}
  >
    <boxStyles>
      <text
        fontSize="15px"
      >
        the name of the force
      </text>
      <TiEdit
        onClick={[Function]}
        size="18"
        style={
          Object {
            "cursor": "pointer",
            "marginLeft": "4px",
            "verticalAlign": "top",
          }
        }
      />
    </boxStyles>
  </textBubbleStyles>
  <boxStyles
    position="relative"
  >
    <Icon
      as="i"
      name="delete"
      onClick={[Function]}
      size="big"
      style={
        Object {
          "color": "#dc2f2f",
          "cursor": "pointer",
          "position": "absolute",
          "right": "250px",
          "top": "calc(50% - 14px)",
        }
      }
    />
    <svg
      height={80}
      viewBox="0 0 250 80"
      width={250}
    >
      <Rectangle
        color="#FFF"
        height={40}
        width={214}
        x={0}
        y={20}
      />
      <Triangle
        color="#FFF"
        height={80}
        width={36}
        x="214,0"
        y="214,80"
        z="250,40"
      />
    </svg>
    <boxStyles
      left={false}
      position="absolute"
      right="36px"
      top="60px"
    >
      <Icon
        as="i"
        disabled={false}
        name="plus"
        onClick={[Function]}
        size="large"
        style={
          Object {
            "cursor": "pointer",
          }
        }
      />
      <text
        fontSize="18px"
      >
        5
      </text>
      <Icon
        as="i"
        disabled={false}
        name="minus"
        onClick={[Function]}
        size="large"
        style={
          Object {
            "cursor": "pointer",
          }
        }
      />
    </boxStyles>
  </boxStyles>
  <Divider
    hidden={true}
  />
</boxStyles>
`;
